### PR TITLE
Add DoReFa Weight Quantizer

### PR DESCRIPTION
**Pull request by Joschua-Conrad**
I mentioned in #620, that I use larq to implement a DoReFa weight quantizer I want to publish with this PR.

The code is added in quantizers.py together with an additional compat reference like it exists for the DoReFa activation quantizer. The new quantizer inherits from the activation quantizer, as it just adds preprocessing to the actual quantize step. In the [Paper](https://arxiv.org/pdf/1606.06160.pdf), section 2.3, one can see in formula (9), that the weight quantizer is defined by using the activation quantizer _quantizek_ together with limiting done with _tanh_ and some range-scaling operations.

I also added unit tests, which check the actual quantization and the gradient. The derivative of _tanh(x)_ is _1/cosh(x)**2_, which makes checking the gradient easy. The quantization can be checked by applying the desired preprocessing and then testing the preprocessed data just like the activation quantizer does. The tests pass.

Please check my docstring and review my comments. If I babble too much, delete the comments or tell me to do it. I could not built the docs, but I hope that my formulas look ok. I also don't know, wether saying
_The interface of this quantizer is the same like the one of DoReFa._
is ok for the docstring of if I should repeat the interface description of the DoReFa activation quantizer. And I don't know wether adding
```
plot-activation
quantizers.DoReFaKernel
```
to the docstring is ok and if the API doc quantizer figure is built automatically, but it fairly looks like it is.

And one last question: the activation quantizer is not included in _TestCommonFunctionality_, so I also did not include my own quantizer there. Is this ok?

Thank you for your time!

### CODE REVIEW

**Comment by jneeven** [comment]
Hi, thanks a lot for this PR! I've been wanting to look at it for a few days now, but unfortunately I'm really busy at the moment and of course I'll need to verify the correctness of your implementation with the paper etc. I'll look at this as soon as I have time do to so, but from a quick glance this looks great 🚀

**Comment by Joschua-Conrad** [comment]
Take your time! Relating the formulas in the paper to my code is probably not that trivial. I hope the comments help.

File: larq/quantizers.py
Comment (lgeiger) [inline larq/quantizers.py:666]: Why is this necessary? I don't see any mention of this in the paper and the linked [reference implementation](https://github.com/tensorpack/tensorpack/blob/5845ead80e3d77c89e2f1e9d5e0691324a8354cc/examples/DoReFa-Net/dorefa.py#L35) also doesn't seem to do this.

TensorFlow will return `inf` for  `x / 0`, which obviously isn't ideal, but I don't think this code would throw a divide by zero error.

File: larq/quantizers.py
Comment (lgeiger) [inline larq/quantizers.py:683]:
```suggestion
        normed = limited / dividend - 0.5
```

File: larq/quantizers.py
Comment (lgeiger) [inline larq/quantizers.py:667]:
```suggestion
        dividend = 2 * tf.math.reduce_max(tf.math.abs(limited))
```

File: larq/quantizers_test.py
Comment (lgeiger) [inline larq/quantizers_test.py:465]:
```suggestion
```

**Review by lgeiger — COMMENTED** [review]
Thanks again for the PR! I made a few initial comments and have some questions about the implementation.

File: larq/quantizers.py
Comment (Joschua-Conrad) [reply larq/quantizers.py]: The suggested change is missing the scaling factor. Commited the corrected suggested change in 426c64e93f029d53b8239774e262dc1ea9bb509c

File: larq/quantizers.py
Comment (Joschua-Conrad) [reply larq/quantizers.py:666]: Sorry when I resolved https://github.com/larq/larq/pull/651#discussion_r620329710, I did not see that you include the scaling factor into the dividend. I suggest keeping all the scaling in the last two code lines, to have the scaling and the reversal scaling all in one place. See 426c64e93f029d53b8239774e262dc1ea9bb509c .

File: larq/quantizers.py
Comment (Joschua-Conrad) [reply larq/quantizers.py:608]: I chose this behavior, because a new class integrates nicely into the automatic creation of the docs and into the unittest framework, where everything is optimized for "one lass per quantizer behavior". I thought about using an instance of the activation quantizer inside the weight quantizer instead of using inheritance, but I am not sure about side effets, I would need to forward arguments to this inner instance and I would need to forward calls to get_config. Using inheritance was the cleanest solution I found.

File: larq/quantizers.py
Comment (Joschua-Conrad) [reply larq/quantizers.py:665]: If this would ocur in one layer, it would output _inf_ for at least one output pixel. The following batch norm layer would calculate the mean of some usual numbers and one _inf_, which is _inf_. The BN layer would apply a _- inf_ operation to all pixels to normalize the mean and the _inf_ value spreads to all pixels. So if this occurs in one spot, it ruins the entire network. Having all weights at exactly zero is anyhow a rare occasion, but having _inf_ or _-inf_ for all outputs is probably not what the user expects. So I added this check. Do you agree?

**Comment by Joschua-Conrad** [comment]
Three questions are open. Please check @lgeiger

File: larq/quantizers.py
Comment (lgeiger) [reply larq/quantizers.py:665]: Makes sense. I looked at the code again and the only way this could happen is when all element of `limited` are exactly zero. This means we can replace `limited / dividend` with [`tf.math.divide_no_nan(limited, dividend)`](https://www.tensorflow.org/api_docs/python/tf/math/divide_no_nan) which returns `0` in cases `dividend == 0` which allows to remove the special handling on line 667 without changing the computation.

Comment (Joschua-Conrad) [reply larq/quantizers.py]: Done in b4a6e29db3a1a6caaa0503d8896395badc14b659

File: larq/quantizers.py
Comment (lgeiger) [inline larq/quantizers.py:617]:
```suggestion
    w_{lim}(w) = \tanh(w)
```

File: larq/quantizers.py
Comment (lgeiger) [inline larq/quantizers.py:625]:
```suggestion
    w_{norm}(w) = \frac{w}{\max(|w|)}
```

File: larq/quantizers.py
Comment (lgeiger) [inline larq/quantizers.py:629]:
```suggestion
    the paper refers to weights being quantized on a numeric range of [-1, 1], while
```

File: larq/quantizers.py
Comment (lgeiger) [inline larq/quantizers.py:630]:
```suggestion
    activations are quantized on the numeric range [0, 1]. `DoReFa` defines the
```

File: larq/quantizers.py
Comment (lgeiger) [inline larq/quantizers.py:632]:
```suggestion
    range of [0, 1], which is why the range needs to be adapted before applying
```

File: larq/quantizers.py
Comment (lgeiger) [inline larq/quantizers.py:639]:
```suggestion
    q(w) = 2 \, quantize_{k}(\frac{w_{norm}\left(w_{lim}\left(w\right)\right)}{2} + \frac{1}{2}) - 1
```

File: larq/quantizers.py
Comment (lgeiger) [inline larq/quantizers.py:643]:
```suggestion
        This quantizer works for weights on the range [-1, 1], which matches the
```

File: larq/quantizers.py
Comment (lgeiger) [inline larq/quantizers.py:645]:
```suggestion
        with a different constraint `clip_value` than the default one.
```

**Review by lgeiger — COMMENTED** [review]
I added a few minor suggestions to improve docs rendering.

**Review by lgeiger — APPROVED** [review]
This is great! Thanks a lot for the PR 🚀 

@larq/core Could we get another pair of eyes to review this before we merge?

**Comment by jneeven** [comment]
> @larq/core Could we get another pair of eyes to review this before we merge?

I hope to look at this early next week!

File: larq/quantizers.py
Comment (jneeven) [reply larq/quantizers.py:608]: For cleanliness of the code I'd prefer to have just a single quantizer class (e.g. like the preliminary version I wrote [here](https://github.com/larq/larq/issues/487#issuecomment-630748936)), but the docs are a good point. I think I would slightly prefer having a base `DoReFa` with `mode="activations"` and then just create a `DoReFaKernel` wrapper that sets `mode="weights"`. No strong opinions though.

File: larq/quantizers.py
Comment (jneeven) [inline larq/quantizers.py:614]:
```suggestion
    with a gradient, which is continuously differentiable itself.
```

File: larq/quantizers.py
Comment (jneeven) [inline larq/quantizers.py:639]: @lgeiger  can we get a render preview somewhere?

File: larq/quantizers.py
Comment (jneeven) [inline larq/quantizers.py:660]:
```suggestion
        limited = tf.math.tanh(inputs)
```

File: larq/quantizers.py
Comment (jneeven) [inline larq/quantizers.py:680]:
```suggestion
        normed = tf.math.divide_no_nan(limited, 2.0 * dividend) + 0.5
```
This keeps the notation closer to the formula in the paper, if you don't mind

File: larq/quantizers.py
Comment (jneeven) [inline larq/quantizers.py:609]: When `k=1`, the DoReFa weight quantizer equals `SteSign`. Is this currently handled correctly?

File: larq/quantizers_test.py
Comment (jneeven) [inline larq/quantizers_test.py:261]: @lgeiger I see this computation was already in the original dorefa unit test, what's going on here? For some reason the index is some complicated `AND` operation?

File: larq/quantizers_test.py
Comment (jneeven) [inline larq/quantizers_test.py:388]: Can you add a case for `k=1` as well?

**Review by jneeven — COMMENTED** [review]
Looks good! Thanks a lot for this. I just have some questions and minor remarks.

File: larq/quantizers.py
Comment (Joschua-Conrad) [reply larq/quantizers.py:608]: This decision affects building the docs and testing, where I do not know the larq project as good as others do. Please agree on one approach and I'll change (or not change) the code accordingly. I took the subclass-way, because it leaves a clear border between activation and weight quantizer without the need for forwarding any calls. Instead using a new keyword for the activation quantizer and some kind of a wrapper for creating the weight quantizer skips usage of inheritance and possible pitfalls.

File: larq/quantizers.py
Comment (Joschua-Conrad) [reply larq/quantizers.py:609]: It is, because the weight quantizer only adds a preprocessor for the activation quantizer.

File: larq/quantizers_test.py
Comment (Joschua-Conrad) [reply larq/quantizers_test.py:261]: Here, ``result`` and ``preprocessed`` have exactly the same shape. One could feed ``preprocessed`` into the activation quantizer and would get exactly ``result``. 

``
(preprocessed > (2 * i - 1) / (2 * n))
& (preprocessed < (2 * i + 1) / (2 * n))
``

creates a boolean index matrix, which marks the position of the values, which fall into the current quantization step bound ``i``. So after applying this matrix to the quantized ``result``, all the returned elements need to have the same value, because they all where inside one quantization step bound in ``preprocessed`` and were quantized to the exact same value.

File: larq/quantizers_test.py
Comment (Joschua-Conrad) [reply larq/quantizers_test.py:389]: Thanky for that remark, this is indeed missing!

Done in 82df95fed124151b278096e81620a6253283737f

This also demonstrates validity of answer in https://github.com/larq/larq/pull/651#discussion_r625002136

Had to increase tolerance of numpy's ``assert_allclose`` by a factor of 10. But this is due to numeric instability. The relative tolerance is 1e-6. The relative tolerance of an 8 bit quantization is 1/256 = 3.9e-3, so actual quantization errors will still be detected.

**Comment by Joschua-Conrad** [comment]
@jneeven Thank your for your review. I added testcases for more than just a 2 bit quantization and accepted your suggestions.

I'm awaiting the following decisions:

- Is the k_bit testcase ok?
- Do we add the different k_bit testcases also to the activation quantizer?
- Do I need to create a ``mode`` argument for the activation quantizer and restructure the implementation of the weight quantizer?

File: larq/quantizers_test.py
Comment (jneeven) [reply larq/quantizers_test.py:389]: Looks great, thanks!

File: larq/quantizers_test.py
Comment (jneeven) [reply larq/quantizers_test.py:261]: Makes sense, thanks!

File: larq/quantizers.py
Comment (lgeiger) [reply larq/quantizers.py:639]: This is done in https://github.com/larq/docs and unfortunately we don't have an automated way to render a preview. However, I manually checked and it renders correctly.

**Review by jneeven — APPROVED** [review]
> Is the k_bit testcase ok?

Looks good to me!

> Do we add the different k_bit testcases also to the activation quantizer?

Would be nice! But we can also do that in a separate PR.

> Do I need to create a mode argument for the activation quantizer and restructure the implementation of the weight quantizer?

This works, so let's leave it like this :+1:

**Comment by Joschua-Conrad** [comment]
Thank you for reviewing!

> > Do we add the different k_bit testcases also to the activation quantizer?
> 
> Would be nice! But we can also do that in a separate PR.

I will create a new PR as soon as this one is merged and I will do so using the updated master branch. If we then find some issue with the multibit testing, we maybe need to edit activation and weight quantizer testing at the same time so I want the weight quantizer multibit testing to be already there.

File: larq/quantizers.py
Comment (lgeiger) [reply larq/quantizers.py:608]: Thinking about this again, we can easily change the docs generation to add autogenerated plots also for quantizers with specific arguments (I am happy to take this on in a follow up PR since it requires changes in some not well documented code paths).

Following the argument from @jneeven above I think I now agree that using a single `DoReFa` class with a `mode` argument that defaults to `"activations"` is probably the more consistent approach within our current API. This might come up again in the future where different quantizers might need different behaviour depending on whether they are used for weights or activations. An example for this would be  int8 quantizers that require different per-channel setting depending on whether they are used for weights or activations. Adding separate classes for each of these would probably increase the API surface too much in my opinion.
I think I am also against exposing a separate `DoReFaKernel` wrapper in this case so this PR would reduce to adding support for a `mode` argument to the existing `DoReFa` class. Would this work for your use case?

Sorry for the back and forth but since this changes the public API I think we want to get it right so we don't have to make any breaking changes later on.

**Review by lgeiger — CHANGES_REQUESTED** [review]

**Comment by lgeiger** [comment]
@Joschua-Conrad Thanks for the fast reply and sorry for the back and forth. Can you take a look at my comment above regarding the public API?

Other than that this looks great and I hope we can merge this soon and make a new release.

**Comment by Joschua-Conrad** [comment]
> @Joschua-Conrad Thanks for the fast reply and sorry for the back and forth. Can you take a look at my comment above regarding the public API?
> 
> Other than that this looks great and I hope we can merge this soon and make a new release.

No prob. Please check my roadmap for the upcoming changes.

@lgeiger @jneeven

File: larq/quantizers.py
Comment (lgeiger) [reply larq/quantizers.py:608]: This sounds great!

File: larq/quantizers_test.py
Comment (lgeiger) [inline larq/quantizers_test.py]: If you still want to test this, I think we can use something like:
```python
("dorefa_kernel_quantizer", functools.partial(lq.quantizers.DoReFa, mode="weights")),
```
Otherwise, I think it is also fine to not add it here.

File: larq/quantizers.py
Comment (lgeiger) [inline larq/quantizers.py:615]:
```suggestion
        ValueError for bad value of `mode`.
```

File: larq/quantizers.py
Comment (lgeiger) [inline larq/quantizers.py:626]: I think using `mode in ("activations", "weights")` is more consistent with the naming Keras uses, since kernel can have multiple meanings depending on the context.

File: larq/quantizers.py
Comment (lgeiger) [inline larq/quantizers.py:627]: Would be great if the error message would also list the available modes. This can be very useful if there is a small typo in the argument that might be easy to overlook.

File: larq/quantizers.py
Comment (lgeiger) [inline larq/quantizers.py:628]: If mode is not already a string, the line above would throw.
```suggestion
        self.mode = mode
```

**Review by lgeiger — CHANGES_REQUESTED** [review]
This is great! some more final comment, before this is good to go 🚀

File: larq/quantizers.py
Comment (Joschua-Conrad) [reply larq/quantizers.py:626]: Done in 363c60cccdaaf69924173ffa8f74a899d71d8c38

File: larq/quantizers.py
Comment (Joschua-Conrad) [reply larq/quantizers.py:627]: List available modes: ef76f89b1e170833b3fd4a429960a984c6decca7
Test invalid mode exceptions: b3fbc421cf4bdd68d546bfc438fa310d70fee217

File: larq/quantizers_test.py
Comment (Joschua-Conrad) [reply larq/quantizers_test.py]: Added testcase in b4bfd473f3ca7878dd10b6122a9d386404f73d89

Added
```
("dorefa_quantizer", functools.partial(lq.quantizers.DoReFa, mode="weights")),
```

because the quantizer string needs to be a valid one, because the test routine passes the quantizer name for some layers, which try to retrieve a valid class for the name string.

Please check!

**Comment by Joschua-Conrad** [comment]
@lgeiger Done. Only one point remains open.

I also added the black conversions and the exception unittests, such that all CI runs should pass now.

File: larq/quantizers.py
Comment (lgeiger) [inline larq/quantizers.py:604]:
```suggestion
        mode: `"activations"` for clipping inputs on [0, 1] range or `"weights"` for
```

File: larq/quantizers.py
Comment (lgeiger) [inline larq/quantizers.py:629]:
```suggestion
                "Valid values are 'activations' and 'weights'."
```

File: larq/quantizers.py
Comment (lgeiger) [inline larq/quantizers.py:660]:
```suggestion
        return tf.math.divide_no_nan(limited, 2.0 * dividend) + 0.5
```

File: larq/quantizers.py
Comment (lgeiger) [inline larq/quantizers.py:672]:
```suggestion
                "Valid values are 'activations' and 'weights'."
```

File: larq/quantizers_test.py
Comment (lgeiger) [inline larq/quantizers_test.py:361]: For the other gradient test we only use `np.vectorize`, because we require an `if/else` statement which is easier to read that way. Since this is not the case here, the native numpy usage is actually preferred. So we can remove this comment.
```suggestion
```

File: larq/quantizers_test.py
Comment (lgeiger) [inline larq/quantizers_test.py:373]:
```suggestion
        expected_gradient = ste_grad if mode == "activations" else tanh_grad
```

**Review by lgeiger — APPROVED** [review]
Thanks for sticking with this PR through the long review. I only have a few minor suggestions that can be directly applied, other than that this is now ready to merge.

Thanks again, for your patience.

**Comment by Joschua-Conrad** [comment]
Thank you for explaining in a nice and patient way what you wanted to be changed! I learned quite a lot about how to do things right in a review process.

**Comment by lgeiger** [comment]
> Thank you for explaining in a nice and patient way what you wanted to be changed! I learned quite a lot about how to do things right in a review process.

Thanks for the feedback, looking forward to the next PR :)

**Comment by lgeiger** [comment]
The changes are now released as part of larq 0.12 and the docs are available at https://docs.larq.dev/larq/api/quantizers/#dorefa

**Pull request merged**
